internal/multistorage/stats/cache/storage_statuses.go

```diff
@@ -0,0 +1,183 @@
+package cache
+
+import (
+	"fmt"
+	"strings"
+	"time"
+)
+
+// Key identifies a storage and is unique within all WAL-G configurations, in contrast to Name, that is unique only
+// within a single configuration. So, Key can be used to identify storages in cache files shared between several WAL-G
+// configurations.
+type Key struct {
+	Name string
+	Hash string
+}
+
+func ParseKey(str string) Key {
+	delim := strings.LastIndex(str, "#")
+	return Key{str[:delim], str[delim+1:]}
+}
+
+func (k Key) String() string {
+	return fmt.Sprintf("%s#%s", k.Name, k.Hash)
+}
+
+// Aliveness measured in arbitrary units and shows how much a storage is alive. When some operation is performed with
+// this storage, the metrica is increased or decreased based on the result of the operation and its weight.
+type Aliveness int
+
+type storageStatus struct {
+	// PotentialAliveness is the maximum Aliveness value, as if all operations with this storage were successful.
+	PotentialAliveness Aliveness `json:"potential_aliveness"`
+
+	// ActualAliveness is the actual Aliveness value, that depends on the success of performed operations.
+	// If it is equal to PotentialAliveness, the storage is fully alive. If it is 0, the storage is fully dead.
+	ActualAliveness Aliveness `json:"actual_aliveness"`
+
+	Updated time.Time `json:"updated"`
+}
+
+const (
+	aliveLimitNumerator   = 95
+	aliveLimitDenominator = 100
+)
+
+func (s storageStatus) alive() bool {
+	return s.ActualAliveness >= s.PotentialAliveness*aliveLimitNumerator/aliveLimitDenominator
+}
+
+func (s storageStatus) applyExplicitCheckResult(alive bool, checkTime time.Time) storageStatus {
+	if alive {
+		return storageStatus{
+			PotentialAliveness: s.PotentialAliveness,
+			ActualAliveness:    s.PotentialAliveness,
+			Updated:            checkTime,
+		}
+	} else {
+		return storageStatus{
+			PotentialAliveness: s.PotentialAliveness,
+			ActualAliveness:    0,
+			Updated:            checkTime,
+		}
+	}
+}
+
+func (s storageStatus) applyOperationResult(alive bool, weight int, checkTime time.Time) storageStatus {
+	if alive {
+		return storageStatus{
+			PotentialAliveness: expMovingAverage(s.PotentialAliveness, weight),
+			ActualAliveness:    expMovingAverage(s.ActualAliveness, weight),
+			Updated:            checkTime,
+		}
+	} else {
+		return storageStatus{
+			PotentialAliveness: expMovingAverage(s.PotentialAliveness, weight),
+			ActualAliveness:    expMovingAverage(s.ActualAliveness, 0),
+			Updated:            checkTime,
+		}
+	}
+}
+
+const (
+	emaAlphaNumerator       = 875
+	emaAlphaDenominator     = 1000
+	ema1MinusAlphaNumerator = emaAlphaDenominator - emaAlphaNumerator
+)
+
+func expMovingAverage(prevAverage Aliveness, newValue int) Aliveness {
```

Comment:
Now, integer arithmetic is used. If using the floating-point one, we can of course use just the 0.875 factor instead of Numerator / Denominator.

Would you like to make the Aliveness float? I was just trying not to use floating-point arithmetic where it's possible because of the bad accuracy and propagated errors.